File: scalechain-transaction/src/test/scala/io/scalechain/blockchain/transaction/OutputOwnershipCodecsSpec.scala
package io.scalechain.blockchain.transaction

import io.scalechain.blockchain.proto.{Hash, WalletTransaction}
import io.scalechain.blockchain.proto.codec.WalletTransactionCodec
import io.scalechain.blockchain.proto.codec.primitive.CodecSuite
import io.scalechain.blockchain.proto.test.ProtoTestData
import io.scalechain.blockchain.script.ScriptParser

class OutputOwnershipCodecsSpec extends CodecSuite with TransactionTestDataTrait with ProtoTestData {

  if (ChainEnvironmentFactory.getActive().isEmpty)
    ChainEnvironmentFactory.create("testnet")

  "OutputOwnershipCodecsSpec" should {

    "roundtrip" in {
      roundtrip(OutputOwnershipCodec.codec, ADDR1.address)
    }
  }

  "publicKeyScript" should {
    "roundtrip" in {
      val expectedScriptOps = ScriptParser.parse( ADDR1.pubKeyScript.lockingScript() )

      val serialized = ParsedPubKeyScriptCodec.serialize(ADDR1.pubKeyScript)

      val actualScriptOps = ParsedPubKeyScriptCodec.parse(serialized)

      actualScriptOps.scriptOps shouldBe expectedScriptOps
    }
  }
}

File: scalechain-transaction/src/test/kotlin/io/scalechain/blockchain/transaction/OutputOwnershipCodecsSpec.kt
package io.scalechain.blockchain.transaction

import io.kotlintest.KTestJUnitRunner
import io.kotlintest.matchers.Matchers
import io.kotlintest.specs.FlatSpec
import io.scalechain.blockchain.proto.Hash
import io.scalechain.blockchain.proto.WalletTransaction
import io.scalechain.blockchain.proto.codec.CodecTestUtil
import io.scalechain.blockchain.proto.codec.WalletTransactionCodec
import io.scalechain.blockchain.proto.test.ProtoTestData
import io.scalechain.blockchain.script.ScriptParser
import io.scalechain.blockchain.transaction.TransactionTestData.ADDR1
import org.junit.runner.RunWith

@RunWith(KTestJUnitRunner::class)
class OutputOwnershipCodecsSpec : FlatSpec(), Matchers, CodecTestUtil, TransactionTestInterface {

  init {
    "OutputOwnershipCodecsSpec" should {
      "roundtrip" {
        roundTrip(OutputOwnershipCodec, ADDR1.address)
      }
    }

    "publicKeyScript" should {
      "roundtrip" {
        val expectedScriptOps = ScriptParser.parse( ADDR1.pubKeyScript.lockingScript() )

        val serialized = ParsedPubKeyScriptCodec.encode(ADDR1.pubKeyScript)

        val actualScriptOps = ParsedPubKeyScriptCodec.decode(serialized)!!

        actualScriptOps.scriptOps shouldBe expectedScriptOps
      }
    }
  }
}

